### PR TITLE
chore(datastore): Use AppSyncGraphQLRequest instead of SimpleGraphQLRequest for sync queries

### PR DESCRIPTION
**Pull request by richardmcclellan**
This is the first step towards adding pagination support in `DataStore`.  We need to use `AppSyncGraphQLRequest` instead of `SimpleGraphQLRequest`, so that we can use the deserialize the response as an `AppSyncPaginatedResult` using `AppSyncPaginatedResultDeserializer`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: core/src/main/java/com/amplifyframework/api/graphql/QueryType.java [hunk 1]
```diff
@@ -27,7 +27,12 @@
     /**
      * GraphQL query list.
      */
-    LIST;
+    LIST,
+
+    /**
+     * GraphQL query sync.
+     */
+    SYNC;
```
Comment (jamesonwilliams) [inline core/src/main/java/com/amplifyframework/api/graphql/QueryType.java:35]: Almost certainly not a comment for this PR, but while I'm here:

I hadn't thought about this at the time, but _get_, _list_, _sync_: these are really verbs from the AppSync domain, aren't they? GraphQL itself has no such pre-defined operation types. Should we aim to move the `QueryType` enum into `aws-api-appsync`?

File: aws-datastore/src/test/resources/delta-sync-request-document-for-post.txt [hunk 2]
```diff
@@ -1,6 +1,9 @@
-query SyncPosts {
-  syncPosts(lastSync: 123123123) {
+query SyncPosts($lastSync: AWSTimestamp) {
```
Comment (jamesonwilliams) [inline aws-datastore/src/test/resources/delta-sync-request-document-for-post.txt:1]: An `AWSTimestamp` in the wild! How cool :-D

File: aws-datastore/src/test/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactoryTest.java [hunk 3]
```diff
@@ -69,10 +73,10 @@ public void validateRequestGenerationForDeltaSync() throws DataStoreException {
     @Test
     public void validateRequestGenerationForPagination() throws DataStoreException {
         final String nextToken = Resources.readAsString("base-sync-request-next-token-value.txt").trim();
-        assertEquals(
-            Resources.readAsString("base-sync-request-paginating-blog-owners.txt"),
-            AppSyncRequestFactory.buildSyncDoc(BlogOwner.class, null, nextToken)
-        );
+        final GraphQLRequest<Iterable<Post>> request = AppSyncRequestFactory
+                .buildSyncRequest(BlogOwner.class, null, nextToken);
```
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactoryTest.java:77]: nit-pick: consider wrapping on the `=`, not on the `.`.

File: aws-datastore/src/test/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactoryTest.java [hunk 4]
```diff
@@ -30,6 +31,8 @@
 
 import java.util.Map;
 
+import edu.emory.mathcs.backport.java.util.Collections;
```
Comment (jamesonwilliams) [inline aws-datastore/src/test/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactoryTest.java:34]: Just `java.util.Collections` here, not the Emory repackaged one.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java [hunk 5]
```diff
@@ -85,70 +84,32 @@ private AppSyncRequestFactory() {}
      * @throws DataStoreException On Failure to inspect
      */
     @NonNull
-    static <T extends Model> String buildSyncDoc(
-            @NonNull final Class<T> modelClass,
+    static <T, M extends Model> AppSyncGraphQLRequest<T> buildSyncRequest(
```
Comment (jamesonwilliams) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java:87]: What are the possible resolutions for `T`, here -- `Iterable<String>`, and importantly, some upcoming type of paginated result?

File: aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java [hunk 6]
```diff
@@ -85,70 +84,32 @@ private AppSyncRequestFactory() {}
...
-        // Outer container, e.g. query SyncBlogPost {
-        doc.append("query Sync").append(capitalizedPluralName).append(" {\n");
-
-        // Inner directive, e.g. syncBlogPosts(
-        doc.append(padBy(++indent)).append("sync").append(capitalizedPluralName);
+            AppSyncGraphQLRequest.Builder builder = AppSyncGraphQLRequest.builder()
+                    .modelClass(modelClass)
+                    .operation(QueryType.SYNC)
+                    .requestOptions(new DataStoreGraphQLRequestOptions())
+                    .responseType(new TypeToken<Iterable<String>>(){}.getType());
```
Comment (jamesonwilliams) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java:98]: You have your little type factory gizmo where you can do like `TypeFactory.of(Iterable.class, String.class)` right? That'd remove the `gson` import from the top of the doc, which would be cool from an encapsulation standpoint.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java (hunk 5 above)
Comment (richardmcclellan) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java:87]: Yep, right now it is `Iterable<String>`, but my next PR will change it to be `PaginatedResult<ModelWithMetadata<M>>`

File: core/src/main/java/com/amplifyframework/api/graphql/QueryType.java (hunk 1 above)
Comment (richardmcclellan) [reply core/src/main/java/com/amplifyframework/api/graphql/QueryType.java:35]: That makes sense, I will move it to `aws-api-appsync`.  I think there's several more things like this that I will also be moving over soon.

File: aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java (hunk 6 above)
Comment (richardmcclellan) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/appsync/AppSyncRequestFactory.java:98]: Updated!

**Comment by richardmcclellan** [comment]
@jamesonwilliams feedback addressed!

**Pull request merged**